### ISSUE TITLE
Stuck client

### ISSUE THREAD

**Issue by ssrlive** [blechschmidt/tun2proxy#28]
I think this is a relatively serious problem. When we run the tun2proxy and then run `ping` command. now sending the ctrl-c command to tun2proxy causes it to exit, which results in the ping command becoming stuck. This problem can be easily reproduced.

![image](https://user-images.githubusercontent.com/30760636/232716143-0bae53f3-6f19-47d4-8d40-be9bdfae3c09.png)

![image](https://user-images.githubusercontent.com/30760636/232716226-3cde4e26-ef4e-42ae-91dc-025ec4877e4d.png)

When watching YouTube videos on an Android phone, it may randomly freeze and require killing the current YouTube client in order to play again.
![image](https://user-images.githubusercontent.com/30760636/232741654-1420a682-4540-49a5-a055-283d86a5a19d.png)

**Comment by blechschmidt** [comment]
Apparently, this is caused by the virtual DNS feature. Does it improve when `DNS_TTL` is set to 1?
https://github.com/blechschmidt/tun2proxy/blob/0c45714a45e7e7ef55ff73e43e5a6a503d2e2bc5/src/virtdns.rs#L10

If that does not fix the problem, disabling virtual DNS and using the `LocalDnsServer` implementation that was previously used with tun2socks should fix it: https://github.com/ShadowsocksR-Live/ssrDroid/blob/6656610910b4f1140c53e33ae00139e339dce43c/core/src/main/java/com/github/shadowsocks/net/LocalDnsServer.kt
In the long term, I want to implement `LocalDnsServer` into tun2proxy itself, so you would just supply the remote DNS server through something like `--dns tcp://1.1.1.1`. ("Native support for proxying DNS over TCP")

**Comment by blechschmidt** [comment]
Please note that it is important to disable virtual DNS for the fix to work, i.e. `with_virtual_dns()` must be removed here:
https://github.com/blechschmidt/tun2proxy/blob/0c45714a45e7e7ef55ff73e43e5a6a503d2e2bc5/src/android.rs#L42

You can see in the screenshots that `ping baidu.com` resolves `baidu.com` to `198.18.0.0`. This is a special IP address from `198.18.0.0/15`, which is not allocated. Virtual DNS uses IPs from this space for IP address to name mapping. While the tun interface is up, these IP addresses can be handled by the virtual DNS feature. When the tun interface is down, these IP addresses are now routed through the default interface and have become meaningless.

1. A client wants to `curl baidu.com`. It first issues a DNS request for `baidu.com`.
2. The DNS request arrives at the tun interface as a UDP packet. It is intercepted and 198.18.0.0 is mapped to `baidu.com`. A DNS reply is generated, telling the client that `baidu.com` is at 198.18.0.0.
3. The client connects to `198.18.0.0:80`. The TCP packets go to the tun interface. The virtual DNS feature recognizes 198.18.0.0 as an IP address used for virtual DNS and looks up the mapping. Instead of sending a SOCKS connection request to 198.18.0.0, it will send a SOCKS request for `baidu.com`, so that actual DNS resolution is handled by the SOCKS proxy.

It's the same procedure for the ping command, except that ping will never work with tun2proxy because ICMP is not supported by SOCKS or HTTP proxies.

A UDP to TCP proxy for DNS should fix this because the DNS request would always return the correct globally routed IP address for `baidu.com`. In that case, an IP address that was resolved while tun2proxy was running would still be valid when tun2proxy is no longer running.

You wrote:
> When watching YouTube videos on an Android phone, it may randomly freeze and require killing the current YouTube client in order to play again.

When exactly does that happen? Does it only happen after you stop tun2proxy?

**Comment by ssrlive** [comment]
When I removed `.with_virtual_dns()`, the situation got even worse, like this:
![Screenshot_20230419_171651](https://user-images.githubusercontent.com/30760636/233030691-5aed39c1-dc31-49de-a6ff-100f1056a7a6.jpg)

**Comment by blechschmidt** [comment]
Sorry for the late reply. I lack the time to investigate this more closely at the moment, but without virtual DNS, you would need a local UDP to TCP DNS forwarder. Since that would need to run on TCP port 53, it would need to be integrated into tun2proxy, so that it can work with Android. If you have the time to implement this, I am happy to accept a PR. It would still be good to know whether this occurs only after tun2proxy has been started and stopped once. This case would hint at a DNS caching issue. If it always occurred, another issue would be more likely.

> I think it's not close incoming socket correctly.
Right now, I'm not sure how not closing incoming sockets would cause this. I can see how virtual DNS responses being cached by e.g. the YouTube app would cause this though, in particular when using CDNs with varying hostnames.

**Comment by ssrlive** [comment]
When I switched the code back to tun2socks, I found that the stuck issue still occurred. So I thought that this problem may not be caused by tun2proxy. I will investigate it again.

**Comment by ssrlive** [comment]
I encountered a strange phenomenon. When I play YouTube videos in the browser, everything is normal and there is no lag. But when using the YouTube app, there is a lag. It’s very difficult to understand.
